File: ports/go-control-plane/manifest.go
package controlplane

import (
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"strconv"
	"strings"
)

// SyntheticModelDescription states what the synthetic benchmark measures. It is
// a bound model for record/chunk state, never an OS RSS claim and never a file
// count: no files are created at any point.
const SyntheticModelDescription = "synthetic-logical-records; no files created"

// FileRecord is one bounded metadata record: where a file sits, how big it is,
// how many lines it carries, and the digest of its bytes.
type FileRecord struct {
	Path   string
	Size   int64
	Lines  int64
	Digest string
}

// Chunk is a deterministic batch of records. Its digest covers the exact
// canonical bytes of its records, so equal chunks are byte-equal across
// languages and runs.
type Chunk struct {
	Index   int
	Records []FileRecord
	Digest  string
	Bytes   int64
	Lines   int64
}

// ManifestSummary folds chunk digests into one root digest plus totals.
type ManifestSummary struct {
	Root       string
	TotalFiles int
	TotalBytes int64
	TotalLines int64
	ChunkCount int
	RootDigest string
}

// RecordStream is a pull iterator over records. Streaming keeps memory bounded:
// a five-million-line repository is a scale target, not a requirement to hold
// five million records in memory.
type RecordStream func() (FileRecord, bool)

// ChunkStream is a pull iterator over chunks.
type ChunkStream func() (*Chunk, bool)

// SyntheticRecords yields logical records modelling a repository of the given
// shape without creating files or line-sized buffers. Each record's digest is
// explicitly a logical digest of its declared geometry, not of any file bytes.
func SyntheticRecords(fileCount, linesPerFile, bytesPerLine int) (RecordStream, error) {
	if fileCount < 0 || linesPerFile < 0 || bytesPerLine <= 0 {
		return nil, errors.New("synthetic dimensions must be nonnegative with positive bytes_per_line")
	}
	next := 0
	return func() (FileRecord, bool) {
		if next >= fileCount {
			return FileRecord{}, false
		}
		path := fmt.Sprintf("synthetic/module-%08d.src", next)
		logical := fmt.Sprintf("synthetic:%s:%d:%d", path, linesPerFile, bytesPerLine)
		sum := sha256.Sum256([]byte(logical))
		next++
		return FileRecord{
			Path:   path,
			Size:   int64(linesPerFile) * int64(bytesPerLine),
			Lines:  int64(linesPerFile),
			Digest: hex.EncodeToString(sum[:]),
		}, true
	}, nil
}

// ChunksFromRecords batches a record stream into bounded chunks whose digests
// are stable across runs with the same bytes and paths.
func ChunksFromRecords(records RecordStream, chunkRecords int) (ChunkStream, error) {
	if chunkRecords <= 0 {
		return nil, errors.New("chunk_records must be positive")
	}
	index := 0
	return func() (*Chunk, bool) {
		bucket := make([]FileRecord, 0, chunkRecords)
		var lineBuf strings.Builder
		for len(bucket) < chunkRecords {
			rec, ok := records()
			if !ok {
				break
			}
			bucket = append(bucket, rec)
			canonicalRecordLine(&lineBuf, rec)
		}
		if len(bucket) == 0 {
			return nil, false
		}
		sum := sha256.Sum256([]byte(lineBuf.String()))
		var bytes, lines int64
		for _, r := range bucket {
			bytes += r.Size
			lines += r.Lines
		}
		chunk := &Chunk{
			Index:   index,
			Records: bucket,
			Digest:  hex.EncodeToString(sum[:]),
			Bytes:   bytes,
			Lines:   lines,
		}
		index++
		return chunk, true
	}, nil
}

// Summarize folds a chunk stream into totals and the root digest: sha256 over
// the concatenated raw chunk-digest bytes, matching the Python scale layer.
func Summarize(root string, chunks ChunkStream) (ManifestSummary, error) {
	rootHash := sha256.New()
	var summary ManifestSummary
	summary.Root = root
	for {
		chunk, ok := chunks()
		if !ok {
			break
		}
		raw, err := hex.DecodeString(chunk.Digest)
		if err != nil {
			return ManifestSummary{}, fmt.Errorf("chunk %d digest is not valid hex: %w", chunk.Index, err)
		}
		rootHash.Write(raw)
		summary.TotalFiles += len(chunk.Records)
		summary.TotalBytes += chunk.Bytes
		summary.TotalLines += chunk.Lines
		summary.ChunkCount++
	}
	summary.RootDigest = hex.EncodeToString(rootHash.Sum(nil))
	return summary, nil
}

// PeakRecordState is the bound-model peak number of records held at once:
// never the whole stream, always at most one chunk.
func PeakRecordState(fileCount, chunkRecords int) int {
	if fileCount < chunkRecords {
		return fileCount
	}
	return chunkRecords
}

// canonicalRecordLine appends the exact bytes CPython's
// json.dumps(asdict(record), sort_keys=True, separators=(",", ":"),
// ensure_ascii=False) + "\n" produces. Hand-rolled because encoding/json
// escapes HTML-significant characters and U+2028/U+2029, which CPython does
// not; parity here means matching the reference encoder, not Go's default.
func canonicalRecordLine(w *strings.Builder, r FileRecord) {
	w.WriteString(`{"digest":"`)
	w.WriteString(r.Digest)
	w.WriteString(`","lines":`)
	w.WriteString(strconv.FormatInt(r.Lines, 10))
	w.WriteString(`,"path":"`)
	canonicalJSONString(w, r.Path)
	w.WriteString(`","size":`)
	w.WriteString(strconv.FormatInt(r.Size, 10))
	w.WriteString("}\n")
}

func canonicalJSONString(w *strings.Builder, s string) {
	for _, r := range s {
		switch {
		case r == '"':
			w.WriteString(`\"`)
		case r == '\\':
			w.WriteString(`\\`)
		case r == '\b':
			w.WriteString(`\b`)
		case r == '\f':
			w.WriteString(`\f`)
		case r == '\n':
			w.WriteString(`\n`)
		case r == '\r':
			w.WriteString(`\r`)
		case r == '\t':
			w.WriteString(`\t`)
		case r < 0x20:
			const hexDigits = "0123456789abcdef"
			w.WriteString(`\u00`)
			w.WriteByte(hexDigits[(r>>4)&0xF])
			w.WriteByte(hexDigits[r&0xF])
		default:
			w.WriteRune(r)
		}
	}
}

File: ports/typescript-edge/src/manifest.ts
/**
 * Streaming deterministic manifest metadata, transcribed from
 * integrations/scale-5m/aurora_scale/manifest.py.
 *
 * The implementation holds at most one bounded chunk of records at a time; a five-million-line
 * repository is therefore a scale target, not a requirement to hold five million records in
 * memory. Synthetic records model that size without creating millions of files - each record's
 * digest is explicitly a logical digest of its declared geometry, not of any file bytes.
 */

import { canonicalJsonString, type CanonicalValue } from "./canonical.js";
import { sha256Hex } from "./digest.js";

/** States what the synthetic benchmark measures: bound-model record/chunk state, never RSS. */
export const SYNTHETIC_MODEL_DESCRIPTION =
  "synthetic-logical-records; no files created";

export interface FileRecord {
  readonly path: string;
  readonly size: number;
  readonly lines: number;
  readonly digest: string;
}

export interface ManifestChunk {
  readonly index: number;
  readonly records: readonly FileRecord[];
  readonly digest: string;
  readonly bytes: number;
  readonly lines: number;
}

export interface ManifestSummary {
  readonly root: string;
  readonly total_files: number;
  readonly total_bytes: number;
  readonly total_lines: number;
  readonly chunk_count: number;
  readonly root_digest: string;
}

export class ManifestError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ManifestError";
  }
}

const encoder = new TextEncoder();

/**
 * Yields logical records for scale tests; it never creates files or line-sized buffers.
 * Dimensions that make no sense throw rather than degrading to an empty stream.
 */
export async function* syntheticRecords(options: {
  fileCount: number;
  linesPerFile: number;
  bytesPerLine?: number;
}): AsyncGenerator<FileRecord> {
  const bytesPerLine = options.bytesPerLine ?? 80;
  if (
    !Number.isSafeInteger(options.fileCount) || options.fileCount < 0 ||
    !Number.isSafeInteger(options.linesPerFile) || options.linesPerFile < 0 ||
    !Number.isSafeInteger(bytesPerLine) || bytesPerLine <= 0
  ) {
    throw new ManifestError(
      "synthetic dimensions must be nonnegative with positive bytesPerLine",
    );
  }
  for (let index = 0; index < options.fileCount; index += 1) {
    const id = String(index).padStart(8, "0");
    const path = `synthetic/module-${id}.src`;
    const logical = `synthetic:${path}:${options.linesPerFile}:${bytesPerLine}`;
    yield {
      path,
      size: options.linesPerFile * bytesPerLine,
      lines: options.linesPerFile,
      digest: await sha256Hex(logical),
    };
  }
}

/**
 * The exact canonical bytes CPython's json.dumps(asdict(record), sort_keys=True,
 * separators=(",", ":"), ensure_ascii=False) + "\n" produces; sort_keys fixes the field order
 * to digest, lines, path, size.
 */
export function canonicalRecordLine(record: FileRecord): string {
  return `${canonicalJsonString(record as unknown as CanonicalValue)}\n`;
}

function recordBytes(record: FileRecord): Uint8Array {
  return encoder.encode(canonicalRecordLine(record));
}

/**
 * Batches a record stream into bounded chunks whose digests are stable across runs given the
 * same records: the chunk digest covers the exact canonical bytes of exactly its records.
 */
export async function* chunksFromRecords(
  records: AsyncIterable<FileRecord>,
  chunkRecords: number,
): AsyncGenerator<ManifestChunk> {
  if (!Number.isSafeInteger(chunkRecords) || chunkRecords <= 0) {
    throw new ManifestError("chunkRecords must be positive");
  }
  let index = 0;
  let bucket: FileRecord[] = [];
  let encodedLines: Uint8Array[] = [];
  let bufferedBytes = 0;
  for await (const record of records) {
    bucket.push(record);
    const line = recordBytes(record);
    encodedLines.push(line);
    bufferedBytes += line.length;
    if (bucket.length === chunkRecords) {
      yield await seal(index, bucket, encodedLines, bufferedBytes);
      index += 1;
      bucket = [];
      encodedLines = [];
      bufferedBytes = 0;
    }
  }
  if (bucket.length > 0) {
    yield await seal(index, bucket, encodedLines, bufferedBytes);
  }
}

async function seal(
  index: number,
  bucket: readonly FileRecord[],
  encodedLines: readonly Uint8Array[],
  total: number,
): Promise<ManifestChunk> {
  const concatenated = new Uint8Array(total);
  let offset = 0;
  for (const line of encodedLines) {
    concatenated.set(line, offset);
    offset += line.length;
  }
  let bytes = 0;
  let lines = 0;
  for (const record of bucket) {
    bytes += record.size;
    lines += record.lines;
  }
  return {
    index,
    records: bucket,
    digest: await sha256Hex(concatenated),
    bytes,
    lines,
  };
}

/**
 * Folds a chunk stream into totals and the root digest: SHA-256 over the concatenated raw
 * chunk-digest bytes, matching the Python layer. Retention is bounded by 32 bytes per chunk -
 * for a 5M-line tree at 1024-record chunks that is about 156 KiB of digest state, not the
 * records themselves.
 */
export async function summarize(
  root: string,
  chunks: AsyncIterable<ManifestChunk>,
): Promise<ManifestSummary> {
  const parts: Uint8Array[] = [];
  let totalFiles = 0;
  let totalBytes = 0;
  let totalLines = 0;
  let chunkCount = 0;
  for await (const chunk of chunks) {
    parts.push(hexToBytes(chunk.digest, chunk.index));
    totalFiles += chunk.records.length;
    totalBytes += chunk.bytes;
    totalLines += chunk.lines;
    chunkCount += 1;
  }
  const all = new Uint8Array(parts.reduce((n, p) => n + p.length, 0));
  let offset = 0;
  for (const part of parts) {
    all.set(part, offset);
    offset += part.length;
  }
  return {
    root,
    total_files: totalFiles,
    total_bytes: totalBytes,
    total_lines: totalLines,
    chunk_count: chunkCount,
    root_digest: await sha256Hex(all),
  };
}

function hexToBytes(hex: string, chunkIndex: number): Uint8Array {
  if (!/^[0-9a-f]*$/.test(hex) || hex.length % 2 !== 0) {
    throw new ManifestError(`chunk ${chunkIndex} digest is not valid lowercase hex`);
  }
  const out = new Uint8Array(hex.length / 2);
  for (let i = 0; i < out.length; i += 1) {
    out[i] = Number.parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  }
  return out;
}

/**
 * The bound-model peak number of records held at once: never the whole stream, always at most
 * one chunk. Tests assert this instead of trusting generator laziness.
 */
export function peakRecordState(fileCount: number, chunkRecords: number): number {
  return Math.min(fileCount, chunkRecords);
}
